### PR TITLE
Use LocalDefId for closures more

### PR DESCRIPTION
**Pull request by camsteffen**

### CODE REVIEW

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 1]
```diff
@@ -341,7 +342,7 @@ pub enum FakeReadCause {
     /// FakeRead for that Place outside the closure, in such a case this option would be
     /// Some(closure_def_id).
     /// Otherwise, the value of the optional DefId will be None.
-    ForMatchedPlace(Option<DefId>),
+    ForMatchedPlace(Option<LocalDefId>),
```
Comment (cjgillot) [inline compiler/rustc_middle/src/mir/syntax.rs:345]: Should we add a comment explaining why this variant can never be decoded from upstream crate?

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 2]
```diff
@@ -1085,8 +1086,8 @@ pub enum AggregateKind<'tcx> {
     /// active field index would identity the field `c`
     Adt(DefId, VariantIdx, SubstsRef<'tcx>, Option<UserTypeAnnotationIndex>, Option<usize>),
 
-    Closure(DefId, SubstsRef<'tcx>),
-    Generator(DefId, SubstsRef<'tcx>, hir::Movability),
+    Closure(LocalDefId, SubstsRef<'tcx>),
```
Comment (cjgillot) [inline compiler/rustc_middle/src/mir/syntax.rs:1089]: Likewise.

File: compiler/rustc_middle/src/mir/syntax.rs (hunk 1 above)
Comment (camsteffen) [reply compiler/rustc_middle/src/mir/syntax.rs:345]: I'm afraid I don't know...is it because TypeckResults is never used for upstream crates?

Comment (cjgillot) [reply compiler/rustc_middle/src/mir/syntax.rs:345]: As far as I can tell, `FakeReadClause` and `AggegateKind` are only used for statements that are removed from MIR between borrowck and encoding to metadata.  As a consequence, they can't appear in metadata, so we can't load a version with a non-local `DefId`.

**Comment by camsteffen** [comment]
Rebased and added comments.

@bors r=cjgillot

**Pull request merged**